File: apps/web/lib/api/memories.ts
import { apiFetch } from "./client";

export interface MemoryItem {
  id: string;
  user_id?: string;
  content: string;
  memory_type: string; // preference, fact, recurring_goal, decision
  importance: number;
  confidence?: number;
  source?: string;
  created_at?: string;
}

export async function listMemories(): Promise<MemoryItem[]> {
  try {
    return await apiFetch<MemoryItem[]>("/api/memories");
  } catch {
    return [];
  }
}

export async function createMemory(data: { content: string; memory_type?: string; importance?: number }): Promise<MemoryItem> {
  return apiFetch<MemoryItem>("/api/memories", {
    method: "POST",
    body: JSON.stringify(data),
  });
}

export async function deleteMemory(id: string): Promise<void> {
  await apiFetch(`/api/memories/${id}`, { method: "DELETE" });
}

export const memoriesApi = {
  listMemories,
  createMemory,
  deleteMemory,
};

File: apps/api/app/api/routes/memories.py
from typing import List
from fastapi import APIRouter, Depends, HTTPException, status

from app.database import User
from app.api.deps import get_current_user
from app.schemas.memory import MemoryCreate, MemorySearchRequest
from app.services.memory_service import memory_service

router = APIRouter(prefix="/memories", tags=["Memories"])


@router.post("", status_code=status.HTTP_201_CREATED)
async def create_memory(
    payload: MemoryCreate,
    user: User = Depends(get_current_user)
):
    """Explicitly save a semantic memory or user preference."""
    mem_id = await memory_service.add_memory(
        user_id=user.id,
        content=payload.content,
        memory_type=payload.memory_type,
        importance=payload.importance
    )
    if not mem_id:
        raise HTTPException(status_code=500, detail="Failed to persist memory")
    return {"id": mem_id, "status": "stored", "content": payload.content}


@router.get("")
async def list_memories(
    limit: int = 50,
    user: User = Depends(get_current_user)
):
    """List semantic memories for current user."""
    return await memory_service.list_memories(user_id=user.id, limit=limit)


@router.post("/search")
async def search_memories(
    payload: MemorySearchRequest,
    user: User = Depends(get_current_user)
):
    """Retrieve semantic memories matching a natural language query."""
    results = await memory_service.search_memories(
        user_id=user.id,
        query=payload.query,
        limit=payload.top_k,
        min_importance=payload.min_importance
    )
    return {"query": payload.query, "results": results, "count": len(results)}


@router.delete("/{memory_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_memory(
    memory_id: str,
    user: User = Depends(get_current_user)
):
    """Delete a specific semantic memory."""
    deleted = await memory_service.delete_memory(memory_id)
    if not deleted:
        raise HTTPException(status_code=404, detail="Memory not found or deletion failed")
